2.5-huffman.ipynb

```diff
@@ -1,25 +1,208 @@
+    "tapleafD = TapLeaf().construct_pk(pubkeyD)\n",
+    "tapleafE = TapLeaf().construct_pk(pubkeyE)\n",
+    "\n",
+    "# Sorted queue: (5, A), (4, B), (3, C), (2, D), (1, E)\n",
+    "# Node DE = parent(D,E)\n",
+    "# Weight of DE = 2 + 1 = 3\n",
+    "nodeDE = Node()\n",
+    "nodeDE.left = tapleafD\n",
+    "nodeDE.right = tapleafE\n",
+    "\n",
+    "# Sorted queue: (5, A), (4, B), (3, C), (3, DE), \n",
+    "# Node CDE = parent(C, DE)\n",
+    "# Weight of CDE = 3 + 3 = 6\n",
+    "nodeCDE = Node()\n",
+    "nodeCDE.left = tapleafC\n",
+    "nodeCDE.right = nodeDE\n",
+    "\n",
+    "# Sorted queue: (6, CDE), (5, A), (4, B)\n",
+    "# Node AB = parent(A,B)\n",
+    "# Weight of AB = 5 + 4 = 9\n",
+    "nodeAB = Node()\n",
+    "nodeAB.left = tapleafA\n",
+    "nodeAB.right = tapleafB\n",
+    "\n",
+    "# Sorted queue: (9, AB), (6, CDE)\n",
+    "# Node ABCDE = parent(AB, CDE)\n",
+    "nodeABCDE = Node()\n",
+    "nodeABCDE.left = nodeAB\n",
+    "nodeABCDE.right = nodeCDE\n",
+    "\n",
+    "# Tree construction\n",
+    "taptree = TapTree()\n",
+    "taptree.root = nodeABCDE\n",
+    "taptree.key = internal_pubkey\n",
+    "\n",
+    "segwit_v1_script, tweak, control_map = taptree.construct()\n",
+    "print(\"Taptree descriptor: {}\\n\".format(taptree.desc))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Part 2: Huffman TapTree Constructor Method\n",
+    "\n",
+    "We reconstruct the same TapTree from the example above using the `TapTree.huffman_constructor()` method. \n",
+    "\n",
+    "* That function takes a list of `(weight,tapleaf)` tuples, where:\n",
+    "    * `weight` is an `int` \n",
+    "    * `tapleaf` is a `TapLeaf` object\n",
+    "   "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Example 2.5.2: Construct a TapTree with the Huffman constructor method\n",
+    "* We compare the resulting TapTree with the one constructed manually in example 2.5.1"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "taptree2 = TapTree()\n",
+    "taptree2.key = internal_pubkey\n",
+    "taptree2.huffman_constructor([(5, tapleafA), (4, tapleafB), (3, tapleafC), (2, tapleafD), (1, tapleafE)])\n",
+    "print(\"Taptree descriptor: {}\\n\".format(taptree2.desc))\n",
+    "\n",
+    "segwit_v1_script2, tweak2, control_map2 = taptree2.construct()\n",
+    "print(\"TapTrees are identical: {}\".format(tweak == tweak2))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### _Programming Exercise 2.5.3:_ Compute height of TapLeafs in TapTree\n",
+    "\n",
+    "Verify that the TapTree constructed in the examples above reflect the structure shown in the image at the beginning of the chapter.\n",
+    "\n",
+    "Reminder: The controlblock of a TapLeaf consists of: \n",
+    "* `[Tapscript Version + Internal Public Key] - 33B`\n",
```

Comment:
I think these two items need more clarity. I am unclear on if you are telling me the sizes or doing some arithmetic.